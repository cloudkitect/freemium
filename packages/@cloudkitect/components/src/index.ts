export * from './ck-certificatemanager/ck-dns-validated-certificate';
export * from './ck-cloudfront/ck-distribution';
export * from './ck-core/ck-vendor-tags';
export * from './ck-core/ck-tag';
export * from './ck-core/ck-utils';
export * from './ck-core/ck-region';
export * from './ck-dynamodb/ck-table-v2';
export * from './ck-ec2/ck-vpc';
export * from './ck-ecr/ck-repository';
export * from './ck-ecs/ck-cluster';
export * from './ck-ecs/ck-fargate-clsuter';
export * from './ck-ecs/ck-fargate-service';
export * from './ck-ecs/ck-fargate-task-definition';
export * from './ck-elasticloadbalancingv2/ck-public-application-loadbalancer';
export * from './ck-s3/ck-bucket';
export * from './ck-stack/ck-stack';
