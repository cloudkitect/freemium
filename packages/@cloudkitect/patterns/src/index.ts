export * from './ck-website/ck-serverless-website';
export * from './ck-container/ck-serverless-container-app';