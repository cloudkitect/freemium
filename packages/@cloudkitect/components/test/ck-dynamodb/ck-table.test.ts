/*
 * Copyright (c) CloudKitect Inc. All Rights Reserved.
 */

import { App } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import {
  AttributeType,
  TableClass,
  TableEncryptionV2,
} from 'aws-cdk-lib/aws-dynamodb';
import { CkAccountType, CkStack } from '../../lib';
import { CkTableV2 } from '../../src';

let app: App;
let stack: CkStack;
const stackProps = {
  ckAccountType: CkAccountType.PRODUCTION,
  ckApplication: 'Test',
  ckCompany: 'CloudKitect',
};

describe('Well Architected DynamoDB Table V2', () => {
  beforeEach(() => {
    app = new App();
    stack = new CkStack(app, 'TestStack', stackProps);
  });

  test('default table configurations', () => {
    new CkTableV2(stack, 'WaTable', {
      partitionKey: { name: 'id', type: AttributeType.STRING },
    });

    const template = Template.fromStack(stack);
    expect(template.toJSON()).toMatchSnapshot();
  });


  test.each`
      actual       | expected | accountType
      ${undefined} | ${true}  | ${CkAccountType.PRODUCTION}
      ${null}      | ${true}  | ${CkAccountType.PRODUCTION}
      ${false}     | ${false} | ${CkAccountType.PRODUCTION}
      ${undefined} | ${false} | ${CkAccountType.DEVELOPMENT}
      ${true}      | ${true}  | ${CkAccountType.DEVELOPMENT}
   `(
    'CC-361: PITR should be $expected when $actual in $accountType',
    ({ actual, expected, accountType }) => {
      const myStack = new CkStack(app, 'MyStack', {
        ...stackProps,
        ckAccountType: accountType,
      });

      new CkTableV2(myStack, 'CkTable', {
        partitionKey: { name: 'id', type: AttributeType.STRING },
        pointInTimeRecovery: actual,
      });

      const template = Template.fromStack(myStack);
      template.hasResourceProperties('AWS::DynamoDB::GlobalTable', {
        Replicas: [
          {
            PointInTimeRecoverySpecification: {
              PointInTimeRecoveryEnabled: expected,
            },
            Region: {
              Ref: 'AWS::Region',
            },
          },
        ],
      });
    },
  );

  test.each`
      actual                                | expected                                | accountType
      ${undefined}                          | ${{ SSEEnabled: true, SSEType: 'KMS' }} | ${CkAccountType.PRODUCTION}
      ${null}                               | ${{ SSEEnabled: true, SSEType: 'KMS' }} | ${CkAccountType.PRODUCTION}
      ${TableEncryptionV2.dynamoOwnedKey()} | ${{ SSEEnabled: false }}                | ${CkAccountType.PRODUCTION}
      ${undefined}                          | ${{ SSEEnabled: false }}                | ${CkAccountType.DEVELOPMENT}
   `(
    'CC-360: encryption should be $expected when $actual in $accountType',
    ({ actual, expected, accountType }) => {
      const myStack = new CkStack(app, 'MyStack', {
        ...stackProps,
        ckAccountType: accountType,
      });

      new CkTableV2(myStack, 'WaTable', {
        partitionKey: { name: 'id', type: AttributeType.STRING },
        encryption: actual,
      });

      const template = Template.fromStack(myStack);
      template.hasResourceProperties('AWS::DynamoDB::GlobalTable', {
        SSESpecification: expected,
      });
    },
  );

  test.each`
      actual       | expected
      ${undefined} | ${true}
      ${null}      | ${true}
      ${false}     | ${false}
   `(
    'CC-365: contributor insights should be $expected when $actual in $accountType',
    ({ actual, expected }) => {
      new CkTableV2(stack, 'CkTable', {
        partitionKey: { name: 'id', type: AttributeType.STRING },
        contributorInsights: actual,
      });

      const template = Template.fromStack(stack);
      template.hasResourceProperties('AWS::DynamoDB::GlobalTable', {
        Replicas: [
          {
            ContributorInsightsSpecification: {
              Enabled: expected,
            },
          },
        ],
      });
    },
  );

  test.each`
      actual                                   | expected                                 | accountType
      ${undefined}                             | ${TableClass.STANDARD}                   | ${CkAccountType.PRODUCTION}
      ${null}                                  | ${TableClass.STANDARD}                   | ${CkAccountType.PRODUCTION}
      ${TableClass.STANDARD_INFREQUENT_ACCESS} | ${TableClass.STANDARD_INFREQUENT_ACCESS} | ${CkAccountType.PRODUCTION}
      ${undefined}                             | ${TableClass.STANDARD_INFREQUENT_ACCESS} | ${CkAccountType.DEVELOPMENT}
   `(
    'CC-202: table class should be $expected when $actual in $accountType',
    ({ actual, expected, accountType }) => {
      const myStack = new CkStack(app, 'MyStack', {
        ...stackProps,
        ckAccountType: accountType,
      });

      new CkTableV2(myStack, 'WaTable', {
        partitionKey: { name: 'id', type: AttributeType.STRING },
        tableClass: actual,
      });

      const template = Template.fromStack(myStack);
      template.hasResourceProperties('AWS::DynamoDB::GlobalTable', {
        Replicas: [
          {
            TableClass: expected,
          },
        ],
      });
    },
  );


});
