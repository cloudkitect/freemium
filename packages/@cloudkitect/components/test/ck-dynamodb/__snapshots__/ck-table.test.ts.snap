// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Well Architected DynamoDB Table V2 default table configurations 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "WaTable2C324A95": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "AttributeDefinitions": [
          {
            "AttributeName": "id",
            "AttributeType": "S",
          },
        ],
        "BillingMode": "PAY_PER_REQUEST",
        "KeySchema": [
          {
            "AttributeName": "id",
            "KeyType": "HASH",
          },
        ],
        "Replicas": [
          {
            "ContributorInsightsSpecification": {
              "Enabled": true,
            },
            "PointInTimeRecoverySpecification": {
              "PointInTimeRecoveryEnabled": true,
            },
            "Region": {
              "Ref": "AWS::Region",
            },
            "TableClass": "STANDARD",
          },
        ],
        "SSESpecification": {
          "SSEEnabled": true,
          "SSEType": "KMS",
        },
      },
      "Type": "AWS::DynamoDB::GlobalTable",
      "UpdateReplacePolicy": "Retain",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
